### DISCUSSION TITLE
Merge conflicts caused damaged xcode file

### DISCUSSION THREAD

**Discussion by gsomaratne** [CS342/.github#23] [Technical Questions] [answered]
### In what area do you have a technical challenge?

Xcode, Simulator & Previews

### Description

I was fixing merge conflicts because of the ongoing pull requests that are being merged into main, and I had to resolve some SwiftLint issues so I did so directly into the github files, and now the build and test doesn't work (it did before the merge conflicts were resovled).

### Reproduction

I committed a few times to change swiftlint issues.

### Expected behavior

I thought once the SwiftLint errors were resolved it would build normally.

### Code of Conduct

- [X] I agree to follow this course's Code of Conduct and Contributing Guidelines

**Comment by PSchmiedmayer** [comment, marked as answer]
Thank you for your question @gsomaratne!

Merging a Xcode project file is not a great experience. Here are a few hints what might have gone wrong and how you might want to address this.

It is essential that you can open the project file in a code editor. I generally use [VS Code](https://code.visualstudio.com) for that and just open to overall application folder (open a folder in VS Code: On the Visual Studio menu bar, choose File > Open > Folder, and then browse to the code location) and then navigate to the project file (e.g., `CS342Application.xcodeproj/project.pbxproj`).

Once you have the project open, you can see a text file with multiple open brackets (`{`) and closed brackets (`};`) that structures the Xcode project. A general hint is to look at the project and see that you can have a closing bracket for all open brackets.

Here are some more details to look at.

## Merge conflict in the `PBXBuildFile / `PBXFileReference` section and general files in files sections of the project:

```
<<<<<<< HEAD
=======
		2F7FC3FD2985DC5D005EDB04 /* File1.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2F7FC3FC2985DC5D005EDB04 /* File1.swift */; };
		2F7FC3FF2985DC89005EDB04 /* Framework1 in Frameworks */ = {isa = PBXBuildFile; productRef = 2F7FC3FE2985DC89005EDB04 /* Framework1 */; };
>>>>>>> main
		2FC8A09B297A315300479145 /* Framework2 in Frameworks */ = {isa = PBXBuildFile; productRef = 2FC8A09A297A315300479145 /* Framework2 */; };
		2FC8A09F297A34C900479145 /* File3.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2FC8A09E297A34C900479145 /* File3.swift */; };
<<<<<<< HEAD
		86FADEAB29886B54008BC807 /* File1.swift in Sources */ = {isa = PBXBuildFile; fileRef = 86FADEAA29886B54008BC807 /* File1.swift */; };
=======
		3648E0E3297F09AD00ADE5C2 /* Framework1 in Frameworks */ = {isa = PBXBuildFile; productRef = 3648E0E2297F09AD00ADE5C2 /* Framework1 */; };
>>>>>>> main
```
Should normally be merged by just accepting both changes, resulting in
```
		2F7FC3FD2985DC5D005EDB04 /* File1.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2F7FC3FC2985DC5D005EDB04 /* File1.swift */; };
		2F7FC3FF2985DC89005EDB04 /* Framework1 in Frameworks */ = {isa = PBXBuildFile; productRef = 2F7FC3FE2985DC89005EDB04 /* Framework1 */; };
		2FC8A09B297A315300479145 /* Framework2 in Frameworks */ = {isa = PBXBuildFile; productRef = 2FC8A09A297A315300479145 /* Framework2 */; };
		2FC8A09F297A34C900479145 /* File3.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2FC8A09E297A34C900479145 /* File3.swift */; };
		86FADEAB29886B54008BC807 /* File1.swift in Sources */ = {isa = PBXBuildFile; fileRef = 86FADEAA29886B54008BC807 /* File1.swift */; };
		3648E0E3297F09AD00ADE5C2 /* Framework1 in Frameworks */ = {isa = PBXBuildFile; productRef = 3648E0E2297F09AD00ADE5C2 /* Framework1 */; };
```


## Merge conflict in the `PBXFrameworksBuildPhase` section
```
		653A254A283387FE005D4D48 /* Frameworks */ = {
			isa = PBXFrameworksBuildPhase;
			buildActionMask = 2147483647;
			files = (
<<<<<<< HEAD
				8625132D297FAF2B004014AE /* Framework1 in Frameworks */,
=======
				75CF1E93297CCD1000E7E71B /* Framework2 in Frameworks */,
>>>>>>> main
				31AD2F0D297FBFBE00880C3F /* Framework3 in Frameworks */,
			);
			runOnlyForDeploymentPostprocessing = 0;
		};
```
Should normally be merged by just accepting both changes, resulting in
```
		653A254A283387FE005D4D48 /* Frameworks */ = {
			isa = PBXFrameworksBuildPhase;
			buildActionMask = 2147483647;
			files = (
				8625132D297FAF2B004014AE /* Framework1 in Frameworks */,
				75CF1E93297CCD1000E7E71B /* Framework2 in Frameworks */,
				31AD2F0D297FBFBE00880C3F /* Framework3 in Frameworks */,
			);
			runOnlyForDeploymentPostprocessing = 0;
		};
```

## Merge conflict in the `XCSwiftPackageProductDependency ` section

This is a tricky one and most often a reason for conflicts breaking the Xcode Project file due to a missing closing bracket combination.

```
/* Begin XCSwiftPackageProductDependency section */
		1A8C31D7297F95E1002D6F2F /* Framework1 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework1;
		};
<<<<<<< HEAD
=======
		2F7FC3FE2985DC89005EDB04 /* Framework2 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework2;
		};
>>>>>>> main
		2FC8A09A297A315300479145 /* Framework3 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework3;
		};
<<<<<<< HEAD
		8625132C297FAF2B004014AE /* Framework4 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework4;
=======
		3648E0E2297F09AD00ADE5C2 /* Framework5 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework5;
		};
		75CF1E92297CCD1000E7E71B /* Framework6 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework6;
>>>>>>> main
		};
		F9FB666F297F14C4008CE0DE /* Framework7 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework7;
		};
/* End XCSwiftPackageProductDependency section */
```

Please be especially aware of the
```
<<<<<<< HEAD
		8625132C297FAF2B004014AE /* Framework4 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework4;
=======
```
part of the conflict. Just accepting both changes would miss a `};` closing bracket combination.

Therefore you need to add this manually and merge the structure like this:
```
/* Begin XCSwiftPackageProductDependency section */
		1A8C31D7297F95E1002D6F2F /* Framework1 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework1;
		};
		2F7FC3FE2985DC89005EDB04 /* Framework2 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework2;
		};
		2FC8A09A297A315300479145 /* Framework3 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework3;
		};
		8625132C297FAF2B004014AE /* Framework4 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework4;
		};
		3648E0E2297F09AD00ADE5C2 /* Framework5 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework5;
		};
		75CF1E92297CCD1000E7E71B /* Framework6 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework6;
		};
		F9FB666F297F14C4008CE0DE /* Framework7 */ = {
			isa = XCSwiftPackageProductDependency;
			productName = Framework7;
		};
/* End XCSwiftPackageProductDependency section */
```

Please check these project files if you might have missed an open or closing bracket and if you might be able to manually fix this by opening the project file @gsomaratne 👍

**Reply by gsomaratne** [reply]
If I made these changes in github how can I undo the different commits from when I resolved the merge conflicts? I also don't know how to open merge conflicts in a code editor.

**Reply by gsomaratne** [reply]
I think I found the bracket error in the project file! However, with the ongoing merges there are even more merge conflicts now. Am I supposed to keep resolving the merge conflicts as they come @PSchmiedmayer ?

**Reply by PSchmiedmayer** [reply]
@gsomaratne Great to hear that you could resolve the merge conflict! Pleas mark the response as an answer if you your problem is solved.

If you have resolved one merge conflict successfully we are happy to help you with future ones. This has been our policy on Assignment #1 this year. I will take a look at your PR 👍

**Reply by gsomaratne** [reply]
Great thank you so much!
